testdata/p4_16_samples/virtual3.p4

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#include <core.p4>
+
+// In arch file
+extern Virtual {
+    Virtual();
+    // abstract methods must be implemented
+    // by the users
+    abstract bit<16> f(in bit<16> ix);
+    void increment();
+    bit<16> total();
+}
+
+// User code
+control c(inout bit<16> p) {
+    bit<16>     local;
+
+    Virtual() cntr = {  // implementation
+        bit<16> f(in bit<16> ix) {  // abstract method implementation
+            return (ix + local);
```

Comment:
@ChrisDodd : I have an issue filed to forbid access to the context in an abstract method. We cannot give a semantics to this code. I don't think we should allow this.

Reply:
Unfortunately we can't do that -- the whole point of abstract methods in per-control instances is to get access to the context -- if we disallow that, we might as well require all externs instances be global.

Reply:
Do you need an annotation here?

Reply:
The @synchronous annotation was added in to the abstract method declaration in the extern above.